File: src/lib/agent/safe-backend.ts
/**
 * SafeBackend wraps any BackendProtocol and intercepts execute() calls
 * to block dangerous shell commands before they reach the shell.
 *
 * This is ENFORCEMENT, not advisory — even if the LLM ignores system
 * prompt instructions, these commands will never execute.
 */

// Patterns that ALWAYS block the command (case-insensitive, checked against normalized command)
const BLOCKED_PATTERNS: RegExp[] = [
  // File/directory deletion
  /\brm\s/i,
  /\brmdir\s/i,
  /\bdel\s/i,
  /\bfind\b.*-delete/i,
  /\bxargs\b.*\brm\b/i,

  // Disk/partition operations
  /\bmkfs\b/i,
  /\bfdisk\b/i,
  /\bdd\s/i,

  // Permission/ownership changes
  /\bchmod\s/i,
  /\bchown\s/i,

  // Process control
  /\bkill\b/i,
  /\bkillall\b/i,
  /\bpkill\b/i,

  // System control
  /\bshutdown\b/i,
  /\breboot\b/i,
  /\bhalt\b/i,
  /\bpoweroff\b/i,

  // Firewall
  /\biptables\b/i,
  /\bufw\b/i,
  /\bfirewall-cmd\b/i,

  // User management
  /\buseradd\b/i,
  /\buserdel\b/i,
  /\bpasswd\b/i,
  /\busermod\b/i,

  // Mount operations
  /\bmount\b/i,
  /\bumount\b/i,

  // Service control
  /\bsystemctl\b/i,
  /\bservice\s/i,

  // Cron modification
  /\bcrontab\s+-[re]/i,

  // Privilege escalation
  /\bsudo\b/i,
  /\bsu\s/i,
  /\bdoas\b/i,

  // Remote access
  /\bssh\b/i,
  /\bscp\b/i,
  /\brsync\b.*@/i,

  // Mutating HTTP
  /\bcurl\b.*-X\s*(POST|PUT|DELETE|PATCH)/i,
  /\bwget\b.*--post/i,

  // Pipe to shell (code execution)
  /\|\s*(bash|sh|zsh|fish)\b/i,
  /\beval\b/i,

  // Destructive git
  /\bgit\s+push\s+--force\b/i,
  /\bgit\s+reset\s+--hard\b/i,
  /\bgit\s+clean\s+-f/i,

  // Container deletion
  /\bdocker\s+(rm|rmi)\b/i,

  // Destructive SQL (in case of shell-piped SQL)
  /\bDROP\s+TABLE\b/i,
  /\bDELETE\s+FROM\b/i,
  /\bTRUNCATE\b/i,

  // Background processes
  /\bnohup\b/i,
  /\bdisown\b/i,

  // Fork bombs
  /:\(\)\s*\{/i,
];

interface ExecuteResult {
  stdout: string;
  stderr: string;
  exitCode: number;
}

/**
 * SafeBackend proxies all methods to the inner backend, but intercepts
 * execute() to check commands against the blocklist.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class SafeBackend {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private inner: any;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(innerBackend: any) {
    this.inner = innerBackend;

    // Proxy all methods from inner backend
    return new Proxy(this, {
      get(target, prop, receiver) {
        // Use SafeBackend's own execute
        if (prop === "execute") {
          return target.execute.bind(target);
        }

        // For everything else, delegate to inner backend
        const innerValue = target.inner[prop];
        if (typeof innerValue === "function") {
          return innerValue.bind(target.inner);
        }
        return innerValue;
      },
    });
  }

  async execute(command: string): Promise<ExecuteResult> {
    const normalized = command.trim();

    for (const pattern of BLOCKED_PATTERNS) {
      if (pattern.test(normalized)) {
        const matchedRule = pattern.source;
        console.warn(
          `[SafeBackend] BLOCKED dangerous command: "${normalized}" (matched: ${matchedRule})`
        );
        return {
          stdout: "",
          stderr: `BLOCKED: This command matches a security rule and cannot be executed. Matched pattern: ${matchedRule}. If you need to perform this operation, ask the user to do it manually.`,
          exitCode: 1,
        };
      }
    }

    // Command is safe — delegate to inner backend
    return this.inner.execute(command);
  }
}

File: python/omnimind_agents/safe_backend.py
from __future__ import annotations

import re
from typing import Any, Dict

BLOCKED_PATTERNS = [
    re.compile(r"\brm\s", re.I),
    re.compile(r"\brmdir\s", re.I),
    re.compile(r"\bdel\s", re.I),
    re.compile(r"\bfind\b.*-delete", re.I),
    re.compile(r"\bxargs\b.*\brm\b", re.I),
    re.compile(r"\bmkfs\b", re.I),
    re.compile(r"\bfdisk\b", re.I),
    re.compile(r"\bdd\s", re.I),
    re.compile(r"\bchmod\s", re.I),
    re.compile(r"\bchown\s", re.I),
    re.compile(r"\bkill\b", re.I),
    re.compile(r"\bkillall\b", re.I),
    re.compile(r"\bpkill\b", re.I),
    re.compile(r"\bshutdown\b", re.I),
    re.compile(r"\breboot\b", re.I),
    re.compile(r"\bhalt\b", re.I),
    re.compile(r"\bpoweroff\b", re.I),
    re.compile(r"\biptables\b", re.I),
    re.compile(r"\bufw\b", re.I),
    re.compile(r"\bfirewall-cmd\b", re.I),
    re.compile(r"\buseradd\b", re.I),
    re.compile(r"\buserdel\b", re.I),
    re.compile(r"\bpasswd\b", re.I),
    re.compile(r"\busermod\b", re.I),
    re.compile(r"\bmount\b", re.I),
    re.compile(r"\bumount\b", re.I),
    re.compile(r"\bsystemctl\b", re.I),
    re.compile(r"\bservice\s", re.I),
    re.compile(r"\bcrontab\s+-[re]", re.I),
    re.compile(r"\bsudo\b", re.I),
    re.compile(r"\bsu\s", re.I),
    re.compile(r"\bdoas\b", re.I),
    re.compile(r"\bssh\b", re.I),
    re.compile(r"\bscp\b", re.I),
    re.compile(r"\brsync\b.*@", re.I),
    re.compile(r"\bcurl\b.*-X\s*(POST|PUT|DELETE|PATCH)", re.I),
    re.compile(r"\bwget\b.*--post", re.I),
    re.compile(r"\|\s*(bash|sh|zsh|fish)\b", re.I),
    re.compile(r"\beval\b", re.I),
    re.compile(r"\bgit\s+push\s+--force\b", re.I),
    re.compile(r"\bgit\s+reset\s+--hard\b", re.I),
    re.compile(r"\bgit\s+clean\s+-f", re.I),
    re.compile(r"\bdocker\s+(rm|rmi)\b", re.I),
    re.compile(r"\bDROP\s+TABLE\b", re.I),
    re.compile(r"\bDELETE\s+FROM\b", re.I),
    re.compile(r"\bTRUNCATE\b", re.I),
    re.compile(r"\bnohup\b", re.I),
    re.compile(r"\bdisown\b", re.I),
    re.compile(r":\(\)\s*\{", re.I),
]


class SafeBackend:
    def __init__(self, inner_backend: Any) -> None:
        self.inner = inner_backend

    def __getattr__(self, item: str) -> Any:
        if item == "execute":
            return self.execute
        return getattr(self.inner, item)

    async def execute(self, command: str) -> Dict[str, Any]:
        normalized = command.strip()
        for pattern in BLOCKED_PATTERNS:
            if pattern.search(normalized):
                matched = pattern.pattern
                return {
                    "stdout": "",
                    "stderr": (
                        "BLOCKED: This command matches a security rule and cannot be executed. "
                        f"Matched pattern: {matched}. If you need to perform this operation, "
                        "ask the user to do it manually."
                    ),
                    "exitCode": 1,
                }

        return await self.inner.execute(command)
